### PR TITLE
Marine ens. var.

### PR DESCRIPTION
**Pull request by guillaumevernieres**
Companion PR to https://github.com/NOAA-EMC/global-workflow/pull/3041

This PR needs an update to `jcb-gdas` (https://github.com/NOAA-EMC/jcb-gdas/pull/46)

### CODE REVIEW

File: test/gw-ci/CMakeLists.txt [hunk 1]
```diff
@@ -66,6 +66,36 @@ endif()
...
+    "gdas_marinebmat"
+    "gdas_ocnanalecen"
+    "gdas_marineanlinit"
+    "gdas_marineanlvar"
+    "gdas_marineanlchkpt"
+    "gdas_marineanlfinal"
+    )
+  add_cycling_tests(${pslot} ${YAML_PATH} ${HOMEgfs} ${RUNTESTS} ${PROJECT_SOURCE_DIR} "${TASK_LIST}")
+
+  # HYB WCDA, low-res
```
Comment (AndrewEichmann-NOAA) [inline test/gw-ci/CMakeLists.txt]: no-so-low-res

Comment (guillaumevernieres) [reply test/gw-ci/CMakeLists.txt]: ?

Comment (AndrewEichmann-NOAA) [reply test/gw-ci/CMakeLists.txt]: You're calling both the `WCDA-HYB-C48mx500` and `WCDA-HYB-C384mx025` cases low res

File: parm/soca/berror/soca_ensb.yaml.j2 [hunk 2]
```diff
@@ -0,0 +1,98 @@
+# Configuration for the recentering and re-balancing of the ensemble members
+geometry:
+  mom6_input_nml: mom_input.nml
+  fields metadata: ./fields_metadata.yaml
+
+date: '{{ MARINE_WINDOW_BEGIN | to_isotime }}'
+
+layers variable: [hocn]
+
+increment variables: [tocn, socn, uocn, vocn, ssh, hocn, cicen, hicen, hsnon]
```
Comment (shlyaeva) [inline parm/soca/berror/soca_ensb.yaml.j2]: does this need an update to var names here and in other places?

File: parm/soca/berror/soca_ensrecenter.yaml [hunk 3]
```diff
@@ -4,104 +4,60 @@ geometry:
...
-
-set increment variables to zero: [ssh]
+increment variables:
+- sea_water_potential_temperature
+- sea_water_salinity
+- eastward_sea_water_velocity
+- northward_sea_water_velocity
+set increment variables to zero:
+- eastward_sea_water_velocity
+- northward_sea_water_velocity
```
Comment (shlyaeva) [inline parm/soca/berror/soca_ensrecenter.yaml:19]: Is the change of variables (compared to develop) here intentional?
For my education, what does `set increment variables to zero` mean and when would one use it?

File: test/testreference/C48mx500_3DVarAOWCDA_3dfgat.ref [hunk 4]
```diff
@@ -1,54 +1,41 @@
-Norm of input parameter StdDev: 4.1916150456516459e+02
+Norm of input parameter StdDev: 1.8173379285780410e+02
 CostJb   : Nonlinear Jb = 0.0000000000000000e+00
-CostJo   : Nonlinear Jo(adt_rads_all) = 3.6538107293180053e+02, nobs = 154056, Jo/n = 2.3717419180804417e-03, err = 3.5902523361909755e+00
-CostJo   : Nonlinear Jo(sst_avhrr_ma_l3u) = 2.2674500789797807e+05, nobs = 80795, Jo/n = 2.8064237625840467e+00, err = 3.1933209964416054e-01
-CostJo   : Nonlinear Jo(sst_avhrr_mb_l3u) = 2.4093341994206450e+05, nobs = 87590, Jo/n = 2.7506955125249970e+00, err = 3.1303879303248278e-01
-CostJo   : Nonlinear Jo(sst_viirs_npp_l3u) = 3.5095100144345453e+05, nobs = 86464, Jo/n = 4.0589262750214488e+00, err = 2.7352638780435701e-01
-CostJo   : Nonlinear Jo(sst_viirs_n20_l3u) = 3.4617882174260722e+05, nobs = 82328, Jo/n = 4.2048734542635220e+00, err = 2.7610924066693887e-01
+CostJo   : Nonlinear Jo(adt_rads_all) = 8.9472206004908480e+02, nobs = 305088, Jo/n = 2.9326688039158695e-03, err = 3.6584285974238564e+00
```
Comment (shlyaeva) [inline test/testreference/C48mx500_3DVarAOWCDA_3dfgat.ref:3]: nobs increased by a factor of two compared to develop, is this intentional?

File: utils/soca/gdas_ens_handler.h [hunk 5]
```diff
@@ -136,19 +132,58 @@ namespace gdasapp {
...
       recenteringIncr.diff(determTraj, ensMeanTraj);
-      eckit::LocalConfiguration sshRecErrOutputConfig(fullConfig, "ssh output.recentering error");
-      recenteringIncr = postProcIncr.setToZero(recenteringIncr);
+      postProcIncr.setToZero(recenteringIncr);
+
+      // Check if we're only re-centering the ensemble fcst around the det.
+      bool recenterOnly(false);
+      if ( fullConfig.has("recentering around deterministic") ) {
+        fullConfig.get("recentering around deterministic", recenterOnly);
+      }
```
Comment (shlyaeva) [inline utils/soca/gdas_ens_handler.h]: Nitpick: something like
```c++
bool recenterOnly = fullConfig.getBool("recentering around deterministic", false);
```
is a bit shorter. No need to change this though.

**Review by shlyaeva — COMMENTED** [review]
A very non-expert review. I also didn't run or test anything, just reviewed the changes.

File: parm/soca/berror/soca_ensb.yaml.j2 (hunk 2 above)
Comment (guillaumevernieres) [reply parm/soca/berror/soca_ensb.yaml.j2]: Oops, I need to delete these files. They were moved to `jcb-gdas`.

File: parm/soca/berror/soca_ensrecenter.yaml (hunk 3 above)
Comment (guillaumevernieres) [reply parm/soca/berror/soca_ensrecenter.yaml:19]: we use it when writing out the mom6 iau increment (for now), and to zero out the background error for selected variables.

File: test/testreference/C48mx500_3DVarAOWCDA_3dfgat.ref (hunk 4 above)
Comment (guillaumevernieres) [reply test/testreference/C48mx500_3DVarAOWCDA_3dfgat.ref:3]: yes, we're adding ssh obs from previous cycles

File: utils/soca/gdas_ens_handler.h (hunk 5 above)
Comment (guillaumevernieres) [reply utils/soca/gdas_ens_handler.h]: looks cleaner, will change.

File: parm/soca/berror/soca_ensrecenter.yaml (hunk 3 above)
Comment (shlyaeva) [reply parm/soca/berror/soca_ensrecenter.yaml:19]: why zero out water velocities in this case?

**Review by shlyaeva — DISMISSED** [review]

File: parm/soca/berror/soca_ensrecenter.yaml (hunk 3 above)
Comment (guillaumevernieres) [reply parm/soca/berror/soca_ensrecenter.yaml:19]: It's never shown to be beneficial so far but we plan to revisit this when running with 80 members. We were also getting increments that were way too large when we were using a "climatological" ensemble.

**Comment by guillaumevernieres** [comment]
> Automated GW GDASApp Testing Results: Machine: hera
> 
> ```
> Start: Mon Dec  2 20:56:32 UTC 2024 on hfe11
> ---------------------------------------------------
> Build:                                  *FAILED*
> Build: Failed at Mon Dec  2 21:14:36 UTC 2024
> Build: see output at /scratch1/NCEPDEV/da/role.jedipara/CI/GDASApp/workflow/PR/1353/global-workflow/sorc/log.build
> ```

Not quite sure why it returned "Build Failed", it's still building.

**Comment by guillaumevernieres** [comment]
ctests started:
ctests-1353

**Comment by guillaumevernieres** [comment]
The new batch of ctests is just not practical. I'm not even sure this is going to run overnight. 
1 - ~we need to add dependencies to the ctests, if a test fail, the subsequent tests shouldn't run~
2 - we should add labels specific to aero/atm/snow and marine. Our work is usually perpendicular to each-others, that pr for example should only run the marine DA stuff
3 - We shouldn't run the forecast after the DA, this will be tested again in the g-w tests ... debatable ...
4 - is it time to stage the model fcst from the 1/2 cycle on hpc's so we don't have to always run the fcst?

I'll create an issue tomorrow with what's above.

**Comment by RussTreadon-NOAA** [comment]
Multiple ctests are failing / will fail because the expanded ctests suite now requires the GSI package (`gsi.x` and `enkf.x`) to be compiled.  For example,
```
 31/134 Test #1964: test_gdasapp_C96C48_hybatmDA_gdas_anal_202112210000 ......................***Failed  210.87 sec
```
failed because `gsi.x` does not exist
```
+ exglobal_atmos_analysis.sh[934]: /bin/cp -p /scratch1/NCEPDEV/da/role.jedipara/CI/GDASApp/workflow/PR/1353/global-workflow/exec/gsi.x /scratch1/NCEPDEV/stmp2/role.jedipara/RUNDIRS/C96C48_hybatmDA/gdas.2021122100/anal.4045548
/bin/cp: cannot stat '/scratch1/NCEPDEV/da/role.jedipara/CI/GDASApp/workflow/PR/1353/global-workflow/exec/gsi.x': No such file or directory
+ exglobal_atmos_analysis.sh[1]: postamble exglobal_atmos_analysis.sh 1733177865 1
+ preamble.sh[70]: set +x
End exglobal_atmos_analysis.sh at 22:18:30 with error code 1 (time elapsed: 00:00:45)
+ JGLOBAL_ATMOS_ANALYSIS[1]: postamble JGLOBAL_ATMOS_ANALYSIS 1733177846 1
```
Script `/scratch1/NCEPDEV/da/role.jedipara/CI/GDASApp/repo/ci/run_gw_ci.sh` builds g-w using
```
rm -rf log.build
./build_all.sh -u &>> log.build
build_status=$?
```
`-u` builds GDASApp.  This was fine when our g-w ctests did not include GSI-based DA. The expanded set of tests now includes GSI-based DA.  We need to add `-g` to build GSI.

I changed `/scratch1/NCEPDEV/da/role.jedipara/CI/GDASApp/repo/ci/run_gw_ci.sh` to read
```
rm -rf log.build
./build_all.sh -ugk &>> log.build
build_status=$?
```
I added `-k` to kill all builds build in case any build fails.

**Comment by guillaumevernieres** [comment]
Damn ... Never ending story.  Of course the one liner change I decide not to test just failed. 
To the reviewers: I'll run the marine DA related "ci" manually on msu tomorrow.

**Comment by guillaumevernieres** [comment]
> Multiple ctests are failing / will fail because the expanded ctests suite now requires the GSI package (`gsi.x` and `enkf.x`) to be compiled. For example,
> 
> ```
>  31/134 Test #1964: test_gdasapp_C96C48_hybatmDA_gdas_anal_202112210000 ......................***Failed  210.87 sec
> ```
> 
> failed because `gsi.x` does not exist
> 
> ```
> + exglobal_atmos_analysis.sh[934]: /bin/cp -p /scratch1/NCEPDEV/da/role.jedipara/CI/GDASApp/workflow/PR/1353/global-workflow/exec/gsi.x /scratch1/NCEPDEV/stmp2/role.jedipara/RUNDIRS/C96C48_hybatmDA/gdas.2021122100/anal.4045548
> /bin/cp: cannot stat '/scratch1/NCEPDEV/da/role.jedipara/CI/GDASApp/workflow/PR/1353/global-workflow/exec/gsi.x': No such file or directory
> + exglobal_atmos_analysis.sh[1]: postamble exglobal_atmos_analysis.sh 1733177865 1
> + preamble.sh[70]: set +x
> End exglobal_atmos_analysis.sh at 22:18:30 with error code 1 (time elapsed: 00:00:45)
> + JGLOBAL_ATMOS_ANALYSIS[1]: postamble JGLOBAL_ATMOS_ANALYSIS 1733177846 1
> ```
> 
> Script `/scratch1/NCEPDEV/da/role.jedipara/CI/GDASApp/repo/ci/run_gw_ci.sh` builds g-w using
> 
> ```
> rm -rf log.build
> ./build_all.sh -u &>> log.build
> build_status=$?
> ```
> 
> `-u` builds GDASApp. This was fine when our g-w ctests did not include GSI-based DA. The expanded set of tests now includes GSI-based DA. We need to add `-g` to build GSI.
> 
> I changed `/scratch1/NCEPDEV/da/role.jedipara/CI/GDASApp/repo/ci/run_gw_ci.sh` to read
> 
> ```
> rm -rf log.build
> ./build_all.sh -ugk &>> log.build
> build_status=$?
> ```
> 
> I added `-k` to kill all builds build in case any build fails.

Thanks for digging into this @RussTreadon-NOAA .

**Comment by guillaumevernieres** [comment]
@DavidNew-NOAA , @RussTreadon-NOAA , @CoryMartin-NOAA , @danholdaway , I'm encapsulating the new ctests inside of cmake options and I'm turning them off by default. For the time being, developers will have to manually turn on/off the tests that are relevant to the code changes. Not great but a simple way to move forward until we have a very low-res atmosphere.

**Comment by DavidNew-NOAA** [comment]
@guillaumevernieres What is the problem with having these tests on by default? If we do encapsulate them in CMake options, then we need a way to turn those options on at build time for nightly testing and PR testing

**Comment by guillaumevernieres** [comment]
> @guillaumevernieres What is the problem with having these tests on by default? If we do encapsulate them in CMake options, then we need a way to turn those options on at build time for nightly testing and PR testing

@DavidNew-NOAA , the tests are too slow, to the point that they are completely useless right now. It wasn't even 1/2 way through this am when I checked.
You can add the options to your cmake build to turn them on for "over-night" testing, but we shouldn't run all these for pr testing.

**Comment by guillaumevernieres** [comment]
testing on my side:
ctests-1353
something fishy with the ens. fcst. rocoto job ... but the ens. forecast was successful. 
Here's the rest:
more-ctests-1353

**Review by shlyaeva — COMMENTED** [review]
I am happy to reapprove all the previous changes (with the caveat of non expert review), but I have no opinion on the latest changes (encapsulating tests in the cmake options). Since it seems that it needs discussion and is not fully related to this PR one option would be to separate that change in a different PR (unless it can be discussed and decided in this PR in a reasonable time).

**Comment by guillaumevernieres** [comment]
> I am happy to reapprove all the previous changes (with the caveat of non expert review), but I have no opinion on the latest changes (encapsulating tests in the cmake options). Since it seems that it needs discussion and is not fully related to this PR one option would be to separate that change in a different PR (unless it can be discussed and decided in this PR in a reasonable time).

Good point. I'll add @DavidNew-NOAA to the PR review, see what he says first. As it is right now, we can't run the ci that we trigger through the github labels.

**Review by DavidNew-NOAA — COMMENTED** [review]
Most is fine by me. I'll just comment on the testing stuff:

1. After some thought, I think it will be easier to only run a subset of tests at the command line rather than with CMake flags. For the ```run_gw_ci.sh``` script in GDASApp [here](https://github.com/NOAA-EMC/GDASApp/blob/4f50d78f872e7c1e29fc4e2c109719e683b4e606/ci/run_gw_ci.sh#L65), we run:

```ctest -j${NTASKS_TESTS} -R gdasapp --output-on-failure &>> log.ctest```

We can just run ```ctest -R C96C48_ufs_hybatmDA``` or ```ctest -R C48mx500_3DVarAOWCDA``` depending on which label one adds to a PR on GitHub.

It will be much easier for me to implement that than turning on those CMake flags at build time, because ```build.sh``` in GDASApp is called by ```build_gdas.sh``` in GW which is called by ```build_all.sh```, so I'll have to had several different options to```build_all.sh```.

2. I was meaning to add a test reference for the C384mx025 marinevar ctest, but I can do it in a separate PR. Up to you.

**Review by AndrewEichmann-NOAA — DISMISSED** [review]
No objections from inspection

**Comment by guillaumevernieres** [comment]
> Most is fine by me. I'll just comment on the testing stuff:
> 
> 1. After some thought, I think it will be easier to only run a subset of tests at the command line rather than with CMake flags. For the `run_gw_ci.sh` script in GDASApp [here](https://github.com/NOAA-EMC/GDASApp/blob/4f50d78f872e7c1e29fc4e2c109719e683b4e606/ci/run_gw_ci.sh#L65), we run:
> 
> `ctest -j${NTASKS_TESTS} -R gdasapp --output-on-failure &>> log.ctest`
> 
> We can just run `ctest -R C96C48_ufs_hybatmDA` or `ctest -R C48mx500_3DVarAOWCDA` depending on which label one adds to a PR on GitHub.
> 
> It will be much easier for me to implement that than turning on those CMake flags at build time, because `build.sh` in GDASApp is called by `build_gdas.sh` in GW which is called by `build_all.sh`, so I'll have to had several different options to`build_all.sh`.
> 
> 2. I was meaning to add a test reference for the C384mx025 marinevar ctest, but I can do it in a separate PR. Up to you.

We're not going to run these test on `hera` for this pr, so I can go back to running the test by default if that helps your nightly testing @DavidNew-NOAA . No need yet for a reference file for the c384.

**Review by KatherineLukens-NOAA — DISMISSED** [review]
I second Andy's comment. No objections here.

**Review by DavidNew-NOAA — APPROVED** [review]
Thanks @guillaumevernieres . I will work on specific CI labels for GitHub testing when Hera is back up

**Review by shlyaeva — APPROVED** [review]

**Pull request merged**